### DISCUSSION TITLE
Is it possible to silence non-verbal parts of an audio?

### DISCUSSION THREAD

**Discussion by orionflame** [openai/whisper#2115] [Q&A]
Hi,

Basically anytime I am not talking it's either silent, some other noise or there is some throat clearing. My audio is noise free so it's quite clear, and I want to keep only the verbal parts, without changing the audio length as its synced to video.

Is there any tool or API that can do this? I tried a few splitter tools online but they failed to remove throat clearing from verbal parts.

I thought maybe I can use the Whisper API to detect the timestamps of where there is speech and silent any other parts. Is that feasible?

It's about 80 hours of audio (~200 files).

Thanks a lot in advance.

**Comment by EtienneAb3d** [comment]
@orionflame 
Remove noise by voice extraction using [Facebook Demucs](https://github.com/facebookresearch/demucs) or [Deezer Spleeter](https://github.com/deezer/spleeter).
See code sample (and other processing tools) here:
https://github.com/EtienneAb3d/WhisperHallu

**Comment by EtienneAb3d** [comment]
@orionflame 
As you suggested, it's possible to identify word positions with timestamps, but you could possibly get some bad word recognitions and/or bad timestamp positions causing the noise removal also damaging the proper voice parts.

**Comment by orionflame** [comment]
Thanks a lot I will try the APIs you mentioned. I didn't have success with other splitters before they just kept the throat clearing parts with vocals.

**Comment by orionflame** [comment]
I just tried Facebook Demucs GUI:
https://www.fosshub.com/Demucs-GUI.html

The throat clearing was still there.

I made a very small sample here:
https://www.dropbox.com/scl/fi/2ambh75jdhfix96wizrcl/voice.mp3?rlkey=1oynf39zlefyyfiuxp45fykth&dl=1

In the middle between 2 statements, there is throat clearing. I have a feeling, throat clearing or coughing is not seen as non-vocal for these models, because they just leave it as is. So I am not sure if they are the right tool for the job.

**Comment by orionflame** [comment]
I tried whisper API which worked for this audio, but another one it didn't remove the throat clearing at all. So I have to see why.

**Comment by orionflame** [comment]
I tried this code that uses the original whisper API on this audio but it didn't silence the throat clearing part on this new audio:
https://paste.ofcode.org/Gc9MUy83K9UHATUHPDVyZ4

https://www.dropbox.com/scl/fi/fi8g0redyz7uzi0fdiil2/voice2.mp3?rlkey=xrmnhrlnaqrzzunwoika0fvj2&dl=1

It worked for the first sample though so I was hopeful.

**Comment by EtienneAb3d** [comment]
@orionflame 
As far as I understand it, the Whisper segmentation is much more a matter of subtitle arrangement than a question of voice part identification.

In my own fast test with your example, here is what should work better for you...
With `WhisperHallu`, call `transcribeOpts` with these options:
```
opts = dict(language='en',word_timestamps=True)
isMusic=True
onlySRT=True
remixFactor=0
```

You should get this:
```
00:00:00,000 --> 00:00:00,340
and

2
00:00:00,340 --> 00:00:00,600
for

3
00:00:00,600 --> 00:00:00,960
curves

4
00:00:00,960 --> 00:00:01,500
they

5
00:00:01,500 --> 00:00:01,860
tell

6
00:00:01,860 --> 00:00:02,040
you

7
00:00:02,040 --> 00:00:02,340
how

8
00:00:02,340 --> 00:00:02,780
quickly

9
00:00:02,780 --> 00:00:03,080
and

10
00:00:03,080 --> 00:00:03,280
in

11
00:00:03,280 --> 00:00:03,460
what

12
00:00:03,460 --> 00:00:03,780
manner

13
00:00:03,780 --> 00:00:04,100
the

14
00:00:04,100 --> 00:00:04,320
curve

15
00:00:04,320 --> 00:00:04,480
is

16
00:00:04,480 --> 00:00:04,880
changing

17
00:00:04,880 --> 00:00:05,140
its

18
00:00:05,140 --> 00:00:05,600
direction.

19
00:00:13,240 --> 00:00:13,880
The

20
00:00:13,880 --> 00:00:14,180
second

21
00:00:14,180 --> 00:00:14,500
order

22
00:00:14,500 --> 00:00:14,920
partial

23
00:00:14,920 --> 00:00:15,340
derivative

24
00:00:15,340 --> 00:00:15,760
is

25
00:00:15,760 --> 00:00:16,060
about

26
00:00:16,060 --> 00:00:16,280
the

27
00:00:16,280 --> 00:00:16,620
curvature

28
00:00:16,620 --> 00:00:17,120
or

29
00:00:17,120 --> 00:00:17,260
the

30
00:00:17,260 --> 00:00:17,480
rate

31
00:00:17,480 --> 00:00:17,840
at

```

Then, remove parts with not-matching word end to next word begin, like:

--> 00:00:**05,600**
direction.

19
**00:00:13,240** -->

You should remove the not-word elements:
![image](https://github.com/openai/whisper/assets/25932245/1af744cb-bc6f-470b-9a09-7d980813797c)

**Comment by orionflame** [comment]
Thanks a lot let me look into this. I will have to adapt my code so that those parts between words are muted basically. Hopefully this will work. Appreciate your help!

**Comment by orionflame** [comment]
Ok I am just looking into it but is it possible to install it with just pip? It shows installing other modules with pip but not the API in question:
https://github.com/EtienneAb3d/WhisperHallu

can I just use?:
pip install WhisperHallu

**Reply by EtienneAb3d** [reply]
@orionflame 
As said on the ReadMe, WhisperHallu is an experimental code.
There are only 3 Python files!

**Reply by EtienneAb3d** [reply]
Be careful, the way it will working depends on the set of modules you will install.

**Comment by orionflame** [comment]
Oh I see, but those 3 files where I can put them? I normally use pip so I don't know where these libraries are stored.

**Reply by EtienneAb3d** [reply]
@orionflame 
Put them at the place your `import` will find them, basically at the same place as your main code. See Code sample in the ReadMe.

**Reply by orionflame** [reply]
Thanks I just tried this:

    import transcribeHallu

    opts = dict(language='en',word_timestamps=True)
    isMusic=True
    onlySRT=True
    remixFactor=0

    model.transcribeOpts(opts)

It complained about:

  File "C:\Users\orionflame\Downloads\WhisperHallu-main\voiceonly.py", line 5, in <module>
    import transcribeHallu
  File "C:\Users\orionflame\Downloads\WhisperHallu-main\transcribeHallu.py", line 43, in <module>
    from demucsWrapper import load_demucs_model
  File "C:\Users\orionflame\Downloads\WhisperHallu-main\demucsWrapper.py", line 3, in <module>
    import demucs
ModuleNotFoundError: No module named 'demucs'

So I need these modules too? In the help file it says only install if you need but I didn't think transcribeOpts calls them.

Also didn't find any info about the transcribeOpts, I guess I am calling it correctly like this?

**Reply by EtienneAb3d** [reply]
As said, WhisperHallu is an experimental code.

You will need some minimal developer skills to understand it and to use it.

For this error case, I need to add a "try..except" to automatically tune the "useDemucs" boolean.
Waiting this, you can change it manually in the code (line 41) to avoid the use of "demucs" if it's not installed.

Your code above should be:
```
model.transcribeOpts(path="PathToYourFile"
  ,opts = dict(language='en',word_timestamps=True)
  ,isMusic=True,onlySRT=True,remixFactor=0)
```

**Reply by orionflame** [reply]
Thanks I will try this.

**Comment by orionflame** [comment]
Now it's trying to find:
Faster installation found, but whisper-medium-ct2/ model not found

I will try to find this model.

**Reply by EtienneAb3d** [reply]
If you properly installed standard Whisper, it should not use Faster Whisper, nor search for it's model.

**Reply by orionflame** [reply]
I just checked your code but there might be bug about which whisper to use. I get printout for both:

Using standard Whisper
Using Faster Whisper
Faster installation found, but whisper-medium-ct2/ model not found

Normally should not continue trying to use faster if standard whisper is found right?

**Reply by EtienneAb3d** [reply]
Answer in the other post (the exchanges would be easier if you would use only one Q/A sequence):
https://github.com/openai/whisper/discussions/2115#discussioncomment-8985643

**Comment by orionflame** [comment]
That's strange. I installed whisper properly I think because I was running the code I posted above that used whisper and it was able to transcribe it.

I will see why that happens/. But in your code you are checking if something is present in whisper and if not you are trying to use faster whisper no?

I only did pip install openai-whisper, nothing else. I downloaded the base.pt model manually and placed it at where it was trying to find it.

**Reply by EtienneAb3d** [reply]
@orionflame 
You should use a [virtualenv](https://virtualenv.pypa.io/en/latest/user_guide.html) if this is not already the case, to be sure of what is installed and detectable while your code is in execution.
If Faster Whisper is detected, it will be used in place of Whisper. But, Faster models should be prepared previously.
Standard Whisper model should be automatically downloaded the first time you transcribe something.

**Reply by orionflame** [reply]
Thanks a lot ok I will do that. I managed to get it running:

modelSize="medium"
# Load the Whisper model
loadModel("0",modelSize=modelSize)

def mute_non_speech(audio_file, model):
    # Load the audio
    audio = whisper.load_audio(audio_file)
    #result = model.transcribe(audio, language="en")

    model.transcribeOpts(path=audio_file
  ,opts = dict(language='en',word_timestamps=True)
  ,isMusic=True,onlySRT=True,remixFactor=0)


But it's trying to use cuda I think. I don't have those installed. I couldn't find any mention of GPU in your project. I changed this line:
device = "cpu" #cuda / cpu

transcribeHallu.py

Is there a way to change it to use CPU? I don't need the fastest speed, just need something that's working.

Using standard Whisper
LOADING: medium GPU:0 BS: 2 PTC=0 TEMP=0
Can't load Whisper model: STD/medium
Attempting to deserialize object on a CUDA device but torch.cuda.is_available() is False. If you are running on a CPU-only machine, please use torch.load with map_location=torch.device('cpu') to map your storages to the CPU.

**Reply by orionflame** [reply]
Ok I managed to set it up cpu by editing your code but now it says:
AttributeError: 'Whisper' object has no attribute 'transcribeOpts'. Did you mean: 'transcribe'?
PS C:\Users\alivs> 

I don't know if I am accessing the model correctly. I was expecting loadModel to return the model but it's using globals.

    model = transcribeHallu.model
    model.transcribeOpts(path=audio_file
  ,opts = dict(language='en',word_timestamps=True)
  ,isMusic=True,onlySRT=True,remixFactor=0)

**Reply by EtienneAb3d** [reply]
@orionflame 
Ok, I introduced a bug when adding the possibility to select between multiple GPU. 
Thanks for this bug report.
Waiting for a fix, change line 118 to enforce CPU, like this:
`
            model = whisper.load_model(modelSize,device=torch.device("cpu"))
`

**Reply by EtienneAb3d** [reply]
`transcribeOpts` is not a method from the Whisper model itself. 

As in the Code Sample, you should simply have:

```
from transcribeHallu import loadModel
from transcribeHallu import transcribeOpts

loadModel("0",modelSize="medium")
result = transcribeOpts(path=audio_file
   ,opts = dict(language='en',word_timestamps=True)
   ,isMusic=True,onlySRT=True,remixFactor=0)

```

**Reply by orionflame** [reply]
Thanks a lot now it seems to be running without errors. I am still waiting for this to finish:
Demucs using device: cpu

Not sure if I shouldn't change demucs to cpu.

Lastly in your first example where you got these:

00:00:00,000 --> 00:00:00,340
and

2
00:00:00,340 --> 00:00:00,600
for

How can I access these? Are these like properties on the result object?

Because in my code I need to use these timings to apply gain to non verbal parts like this:

        start = int(segment["start"] * 1000)  # Convert to milliseconds
        end = int(segment["end"] * 1000)
        
        non_verbal = original_audio[last_pos:start].apply_gain(-100)  # Decrease volume significantly

So not sure if I can get these timings and loop over them like segments?

I just got the same print out as you for the timing but after like 2 mins. I assume demucs on cpu is really slow as this is just like a 5 second audio file.

**Reply by EtienneAb3d** [reply]
@orionflame 
This code was designed to produce this SRT content.
You now have 2 possibilities:

1. re-parse this SRT to get it as a usable Python structure (possibly there are some libraries to do this)
2. adapt de code to build this Python structure in place of the SRT content (this SRT is produced by a Whisper code from line [543](https://github.com/EtienneAb3d/WhisperHallu/blob/9d6b0abf3f7a13e6f54347fafa2128369e72e4fb/transcribeHallu.py#L543C1-L543C48))

**Reply by orionflame** [reply]
Thanks a lot that makes sense now. Ok I will do that and use the output to mute relevant parts.

**Reply by orionflame** [reply]
Also after running your code I get a lot of wav, dur, log, bass, etc files. Is it possible to disable this since I don't need these? Also the subtitles output is printed in the console. Is that also possible to disable? I am just asking before I go in and have to change a lot of things in your code, maybe there are some options that control these.

**Reply by EtienneAb3d** [reply]
@orionflame 

Since all temporary files are obtained adding suffixes to the original, it's easy to delete all these temporary files with a single remove command after each processing. For example, if your files are all ".wav":
`rm *.wav.WAV.*
`

The subtitle output is in the `transcribeOpts` method result. Simply use it.

As far as I understand what you want to do, and your developer skills, I don't think you should change the code. You should rather re-parse the SRT result. You will then also keep available all the WhisperHallu possible configurations/options, like using an other transcriber, or an additional filtering.

**Reply by orionflame** [reply]
But this SRT output is only printed right? Is it also saved on disk?

Before I was commenting out this line but then I didn't get a print out:
save_audio(pathVAD,collect_chunks(speech_timestamps, wav), sampling_rate=SAMPLING_RATE)

and no files either. That's why I was confused. I will use your rm idea.

**Reply by EtienneAb3d** [reply]
@orionflame 
I don't want to hurt you, but, I won't have the time to help you write the code line by line, for so elementary questions.
The SRT value is in `result`, you may simply print it, or do whatever you want with it:
```
result = transcribeOpts(path=audio_file
   ,opts = dict(language='en',word_timestamps=True)
   ,isMusic=True,onlySRT=True,remixFactor=0)
print(result)

```

**Reply by orionflame** [reply]
No worries. I was just trying to clean up the code because just calling the transcribe function it's printing the entire result. IMO this is not intuitive behaviour. Commenting out this line fixes that:

print("TRANS="+result["text"],flush=True)

**Reply by orionflame** [reply]
Thanks a lot for all your help. I wrote the code that basically mutes between verbal parts and it did work in the sample voice files I posted here. But in the first 5 min audio file I tested, I noticed it's removing actual words or parts of them in some places.

I made a short sample here that has the problem areas:
https://www.dropbox.com/scl/fi/kotmse874x4rsi86kr8f8/voice3.mp3?rlkey=l5m56g5axort1ru70goo3rvch&dl=1

I assume the approach you suggested relies on the transcriber perfectly understanding what's being said? Because there are some names that I imagine the transcriber will have problems matching. For example in the name I spoke "Keenan" it cropped it partially in one of the instances.

Now that the solution is fully implemented, I am trying to see if it can indeed be used perfectly.

**Reply by EtienneAb3d** [reply]
Try to use "large-v2" model in place of "medium" one.

**Reply by EtienneAb3d** [reply]
I just heard your example. The word "Keenan" is partially repeated 3 times. 
It's a known behaviour from Whisper models to remove such (supposed bad) repetitions.

In my own fast test with large-v2, there is no real in-between word interval (thus, with the algo discussed above, no text removed), but and empty text for the first "Keen" (= a segment without word highlighting in the original SRT produced by the Whisper code).

**Reply by orionflame** [reply]
Thanks a lot that makes sense. I will try the large model. Basically this is the audio stream for a tutorial. So first I am trying to remove the non verbal parts then my goal was to remove the retakes. Can whisper be used to do this?

Most of the time it could be partial sentences repeated multiple times until a full sentence is formed. But it's always the last instance that should be kept. That's a much bigger problem which I don't know how to solve yet.

Sorry in your last sentence are you saying it successfully transcribes the first "Keen" part?

Also if you check the first part when I say "Pretty" that was cropped in the medium model also. The p is so short, in the original it was normal.

Or do you think I should look into other VAD models like Pyannote? Because I was reading some people saying whisper sometimes misses some words.

**Reply by EtienneAb3d** [reply]
@orionflame 
Certainly a lot of things may be experimented that will bring you with a large part of the solution working quite properly in a lot of cases. But, as far as I know the current state-of-the-art, you won't succeed in building a solution that will ensure it won't damage something that should be kept in the audio. If current timestamp tools are enough precise to get a colour display on subtitles humanly acceptable, they are far too much imprecise for such a sound filtering task.

To remove bad repetitions, you may try to ask ChatGPT to remove them from the SRT (I would use segment-level rather than word-level timestamps). For this, you may use ChatMate. It should be easy to adapt the provided translation example.
https://github.com/EtienneAb3d/ChatMate

**Comment by EtienneAb3d** [comment]
@orionflame 

You may experiment with [whisper-at](https://github.com/YuanGongND/whisper-at).

Combining the text time ranges to know where a text is recognized, with tags explaining what it is suppose to be, using a small time resolution like 1.6s (not too small, to get something in each part, and improve the result quality), you could possibly obtain something very interesting.

Here is the result using `tiny.en` model on your example 2:
![image](https://github.com/openai/whisper/assets/25932245/bc703e84-2483-4756-a797-e6939a856bbe)

You see that, from 6.0s to 13.6s there is no text recognized, while this part is tagged with some "Cough, Throat clearing, Animal" tags that confirm the fact there is something not interesting in this part.

It would remove this part of your audio:
![image](https://github.com/openai/whisper/assets/25932245/b28d85ef-9e69-4901-aa33-4ff5c1fc12ca)

On your example 1, it less an evidence on what should be removed because there isn't a good match between text ranges and unwanted tags. But you could perhaps get a much better result after having applied a noise removal on the sound file.

![image](https://github.com/openai/whisper/assets/25932245/43662e53-1e7d-4921-8fc9-961af78ec3b2)

**Reply by orionflame** [reply]
Thanks a lot I will experiment with this tool today. Although this audio file I already applied noise removal using audacity. Normally it has a bit of background noise.

**Reply by orionflame** [reply]
Also I am getting this text for the last audio:

0.0s-6.9s:  pretty much everything you could want that occur around the normal vector not
6.9s-13.3s:  along it. Keenan Crane is one of the leading
13.3s-17.2s:  researchers in computational geometry.

It is doing de-duplication for the text? I thought maybe I could use that to remove the retakes but it just has a large range like 6.9s-13.3s.

I am surprised it was able to perfectly transcribe keenan crane though but also throat clearing in the audio you tested.

In any case the last audio didnt tag the throat clearing so I got these:

0.0s-1.6s: Speech, Narration, monologue, Speech synthesizer, Clicking, Male speech, man speaking
1.6s-3.2s: Speech, Narration, monologue, Speech synthesizer, Clicking, Male speech, man speaking
3.2s-4.8s: Speech, Inside, small room, Clicking, Speech synthesizer, Narration, monologue
4.8s-6.4s: Speech, Narration, monologue, Speech synthesizer, Male speech, man speaking
6.4s-8.0s: Speech, Narration, monologue, Clicking, Speech synthesizer, Inside, small room
8.0s-9.6s: Speech, Clicking, Inside, small room
9.6s-11.2s: Speech, Clicking, Inside, small room, Narration, monologue, Male speech, man speaking
11.2s-12.8s: Speech, Speech synthesizer
12.8s-14.4s: Sine wave
14.4s-16.0s: Sine wave, Hum, Chime, White noise, Boiling



In any case when using tags to mute the non speech, what would you recommend? Because even throat clearing is tagged with speech in the same line no? I thought maybe there would be a speech tag that's used only when someone is speaking.

**Reply by EtienneAb3d** [reply]
You have to discuss this with the tool authors.

**Reply by orionflame** [reply]
Also how long it takes you to transcribe my audio sample file? On my system I have to wait more than 30 mins. Is this not too slow? Because it's only 12s audio. I was using whisper before, and it was able to create subtitles so much faster like orders of magnitude for 20 hours of audio. I was using stable whisper because regular one had incorrect timings I think:
https://github.com/openai/whisper/discussions/435

I know you also support fast whisper which I want to try today but that would only give me like 4x speed up right?

Should I just switch to gpu or use another whisper that can be drop in replacement?

Because I want to add some padding to the muting so that the cropping issue is gone. There is always some gap between coughs and speech.

I will still try whisper at today but will have to resolve this first:
https://github.com/YuanGongND/whisper-at/issues/26

**Reply by EtienneAb3d** [reply]
Of course, you can do some basic tests using CPU.
But, such models (especially the larger ones) need GPU to be efficient.
